cmd/mt-whisper-importer-reader/conversion.go

```diff
@@ -0,0 +1,209 @@
+					adjustedPoints[m][p.Timestamp] = p.Value
+				}
+			}
+		} else {
+			methods := []string{c.method}
+			if c.method == "avg" && retIdx > 0 {
+				methods = []string{"sum", "cnt"}
+			}
+			for m, points := range decResolution(in, methods, arch.SecondsPerPoint, spp) {
+				for _, p := range points {
+					adjustedPoints[m][p.Timestamp] = p.Value
+				}
+			}
+		}
+
+	}
+
+	// merge the results that are keyed by timestamp into a slice of points
+	for m, p := range adjustedPoints {
+		for t, v := range p {
+			res[m] = append(res[m], whisper.Point{Timestamp: t, Value: v})
+		}
+		res[m] = sortPoints(res[m])
+	}
+
+	return res
+}
+
+// increase resolution of given points according to defined specs
+func incResolution(points []whisper.Point, method string, inRes, outRes, rawRes uint32) map[string][]whisper.Point {
+	out := make(map[string][]whisper.Point)
+	aggFactor := float64(outRes) / float64(rawRes)
+	for _, inPoint := range points {
+		if inPoint.Timestamp == 0 {
+			continue
+		}
+
+		rangeEnd := inPoint.Timestamp - (inPoint.Timestamp % outRes)
+
+		var outPoints []whisper.Point
+		for ts := rangeEnd; ts > inPoint.Timestamp-inRes; ts = ts - outRes {
+			outPoints = append(outPoints, whisper.Point{Timestamp: ts})
+		}
+
+		for _, outPoint := range outPoints {
+			if method == "cnt" || method == "sum" {
+				outPoint.Value = inPoint.Value / float64(len(outPoints))
+			} else if method == "fakeavg" {
+				outPoint.Value = inPoint.Value * aggFactor
+			} else {
+				outPoint.Value = inPoint.Value
+			}
+
+			if method == "fakeavg" {
+				out["sum"] = append(out["sum"], outPoint)
+				out["cnt"] = append(out["cnt"], whisper.Point{Timestamp: outPoint.Timestamp, Value: aggFactor})
+			} else {
+				out[method] = append(out[method], outPoint)
+			}
+		}
+	}
+	for m := range out {
+		out[m] = sortPoints(out[m])
+	}
+	return out
+}
+
+func decResolution(points []whisper.Point, methods []string, inRes, outRes uint32) map[string][]whisper.Point {
+	out := make(map[string][]whisper.Point)
+	agg := mdata.NewAggregation()
+	currentBoundary := uint32(0)
+
+	flush := func() {
+		if agg.Cnt == 0 {
+			return
+		}
+
+		for _, method := range methods {
+			var value float64
+			switch method {
+			case "min":
+				value = agg.Min
+			case "max":
+				value = agg.Max
+			case "sum":
+				value = agg.Sum
+			case "cnt":
+				value = agg.Cnt
+			case "lst":
+				value = agg.Lst
+			case "avg":
+				value = agg.Sum / agg.Cnt
+			default:
+				return
+			}
+			out[method] = append(out[method], whisper.Point{
+				Timestamp: currentBoundary,
+				Value:     value,
+			})
+		}
+		agg.Reset()
+	}
+
+	for _, inPoint := range sortPoints(points) {
+		if inPoint.Timestamp == 0 {
+			continue
+		}
+		boundary := mdata.AggBoundary(inPoint.Timestamp, outRes)
+
+		if boundary == currentBoundary {
+			agg.Add(inPoint.Value)
+			if inPoint.Timestamp == boundary {
+				flush()
+			}
+		} else {
+			flush()
+			currentBoundary = boundary
+			agg.Add(inPoint.Value)
+		}
+	}
```

Comment:
should we  make sure to flush any unflushed points? can you add a comment as to why (not)

Reply:
I'll mention that in a comment too:
Generally we only want to write aggregated points that have been "finished", as in we have a received a point with a `ts` which is `>=` the highest `ts` that would be factored into this point, in any other case the result would likely be wrong because we can't predict the future.